src/stan_math_backend/Lower_functions.ml

```diff
@@ -204,141 +189,98 @@ let lower_fun_def (functors : (string, struct_defn) Hashtbl.t)
+  let fd =
+    match Option.map ~f:(lower_fun_body fdargs fdsuffix) fdbody with
+    | Some body -> almost_fn ~body ()
+    | None -> almost_fn () in
+  let register_functor variadic_fun_type =
+    let suffix = Lower_expr.functor_suffix_select variadic_fun_type in
+    let functor_name = fdname ^ suffix in
+    let struct_template =
+      match (fdsuffix, variadic_fun_type) with
+      | FnLpdf _, ReduceSum -> Some (Bool "propto__")
+      | _ -> None in
+    let arg_templates, templated_args =
+      template_parameter_and_arg_names false variadic_fun_type in
+    let cpp_args = cpp_arg_gen templated_args variadic_fun_type in
+    let defn_template =
+      match fdsuffix with
+      | FnLpdf _ | FnTarget -> [TemplateType "propto__"]
+      | _ -> [] in
+    let defn_args =
+      List.map ~f:Exprs.to_var
+        (List.map ~f:snd templated_args @ extra_arg_names @ ["pstream__"]) in
+    let defn_args =
+      match (variadic_fun_type, defn_args) with
+      | ReduceSum, slice :: start :: end_ :: rest ->
+          slice :: plus_one start :: plus_one end_ :: rest
+      | _ -> defn_args in
+    let defn_body =
+      [Return (Some (Exprs.templated_fun_call fdname defn_template defn_args))]
+    in
+    let functor_decl =
+      make_fun_defn
+        ~templates_init:([arg_templates], true)
+        ~name:"operator()"
+        ~return_type:(lower_returntype fdargs fdrt)
+        ~args:cpp_args ~cv_qualifiers:[Const] ~body:defn_body () in
+    make_struct_defn ~param:struct_template ~name:functor_name
+      ~body:[FunDef functor_decl] () in
+  (fd, functors |> List.map ~f:register_functor)
 
-let is_fun_used_with_reduce_sum (p : Program.Numbered.t) =
-  let rec find_functors_expr accum Expr.Fixed.{pattern; _} =
-    String.Set.union accum
-      ( match pattern with
-      | FunApp (StanLib (x, FnPlain, _), {pattern= Var f; _} :: _)
-        when Stan_math_signatures.is_reduce_sum_fn x ->
-          String.Set.of_list [Utils.stdlib_distribution_name f]
-      | x -> Expr.Fixed.Pattern.fold find_functors_expr accum x ) in
-  let rec find_functors_stmt accum stmt =
-    Stmt.Fixed.(
-      Pattern.fold find_functors_expr find_functors_stmt accum stmt.pattern)
-  in
-  Program.fold find_functors_expr find_functors_stmt String.Set.empty p
-
-let get_variadic_requirements (p : Program.Numbered.t) =
-  let rec find_functors_expr accum Expr.Fixed.{pattern; _} =
-    match pattern with
-    | FunApp (StanLib (x, FnPlain, _), {pattern= Var f; _} :: _) -> (
-      match
-        Hashtbl.find Stan_math_signatures.stan_math_variadic_signatures x
-      with
-      | Some {required_fn_args; _} ->
-          Map.add_multi accum
-            ~key:(Utils.stdlib_distribution_name f)
-            ~data:(List.length required_fn_args)
-      | _ -> Expr.Fixed.Pattern.fold find_functors_expr accum pattern )
-    | _ -> Expr.Fixed.Pattern.fold find_functors_expr accum pattern in
+let get_functor_requirements (p : Program.Numbered.t) =
```

Comment:
I really appreciate how much more general this is than the prior code. Am I correct that as a side effect, we no longer generate any functor if the function is never used in a higher order function?

Reply:
That's correct.